src/detect-mqtt-connack-sessionpresent.c

```diff
@@ -0,0 +1,320 @@
+                               Flow *f, uint8_t flags, void *state,
+                               void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+{
+    const bool *de = (const bool *)ctx;
+    bool value = false;
+
+    if (!de)
+        return 0;
+
+    if (rs_mqtt_tx_get_connack_sessionpresent(txv, &value) ==0 ) {
+        return 0;
+    }
+    if (value != *de) {
+        return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * \internal
+ * \brief This function is used to parse options passed via mqtt.connack.session_present: keyword
+ *
+ * \param rawstr Pointer to the user provided options
+ *
+ * \retval de pointer to DetectMQTTConnackSessionPresentData on success
+ * \retval NULL on failure
+ */
+static bool *DetectMQTTConnackSessionPresentParse(const char *rawstr)
+{
+    bool *de = NULL;
+#define MAX_SUBSTRINGS 30
+    int ret = 0;
+    int ov[MAX_SUBSTRINGS];
+
+    ret = DetectParsePcreExec(&parse_regex, rawstr, 0, 0, ov, MAX_SUBSTRINGS);
+    if (ret < 1) {
+        SCLogError(SC_ERR_PCRE_MATCH, "invalid session_present flag definition: %s", rawstr);
+        return NULL;
+    }
+
+    de = SCMalloc(sizeof(bool));
+    if (unlikely(de == NULL))
+        return NULL;
+    *de = false;
+
+    if (strcmp(rawstr, "yes") == 0) {
+        *de = true;
+    } else if (strcmp(rawstr, "true") == 0) {
+        *de = true;
+    } else if (strcmp(rawstr, "no") == 0) {
+        *de = false;
+    } else if (strcmp(rawstr, "false") == 0) {
+        *de = false;
+    } else {
+        SCLogError(SC_ERR_UNKNOWN_VALUE, "invalid session_present flag definition: %s", rawstr);
+        goto error;
+    }
+
+    return de;
+
+error:
+    SCFree(de);
+    return NULL;
+}
+
+/**
+ * \internal
+ * \brief this function is used to add the parsed type query into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param rawstr pointer to the user provided options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectMQTTConnackSessionPresentSetup (DetectEngineCtx *de_ctx, Signature *s, const char *rawstr)
+{
+    bool *de = NULL;
+    SigMatch *sm = NULL;
+
+    de = DetectMQTTConnackSessionPresentParse(rawstr);
+    if (de == NULL)
+        goto error;
+
+    sm = SigMatchAlloc();
+    if (sm == NULL)
+        goto error;
+
+    sm->type = DETECT_AL_MQTT_CONNACK_SESSION_PRESENT;
+    sm->ctx = (SigMatchCtx *)de;
+
+    SigMatchAppendSMToList(s, sm, mqtt_connack_session_present_id);
+
+    return 0;
+
+error:
+    if (de) SCFree(de);
+    if (sm) SCFree(sm);
+    return -1;
+}
+
+/**
+ * \internal
+ * \brief this function will free memory associated with DetectMQTTConnackSessionPresentData
+ *
+ * \param de pointer to DetectMQTTConnackSessionPresentData
+ */
+void DetectMQTTConnackSessionPresentFree(DetectEngineCtx *de_ctx, void *de_ptr)
+{
+    bool *de = (bool *)de_ptr;
+    if(de) SCFree(de);
```

Comment:
I do not think you need to test for NULL as `SCFree` does it already

Reply:
Ah, this is probably just a habit :) Will keep this in mind.

Reply:
That is a good habit ;-)